### ISSUE TITLE
Consider testing enhancements for `filelog` receiver

### ISSUE THREAD

**Issue by ChrsMark** [open-telemetry/opentelemetry-collector-contrib#32001]
### Component(s)

receiver/filelog

### Describe the issue you're reporting

As of today the OTel Collector can handle log collection without loss or duplication after restarts.
This is possible by using the [storage extension](https://github.com/open-telemetry/opentelemetry-collector-contrib/tree/main/extension/storage/filestorage) to store the receiver's state and using a [persistent-queue](https://github.com/open-telemetry/opentelemetry-collector/blob/main/exporter/exporterhelper/README.md#persistent-queue) to persist the exporter's state.
This is documented at the [offset tracking](https://github.com/open-telemetry/opentelemetry-collector-contrib/blob/main/receiver/filelogreceiver/README.md#offset-tracking) section and the [fault-tolerant-logs-collection](https://github.com/open-telemetry/opentelemetry-collector-contrib/blob/main/examples/fault-tolerant-logs-collection/README.md) example.

However there are cases where it has been reported data loss like at https://github.com/open-telemetry/opentelemetry-collector-contrib/issues/31074, in which is not quite straightforward from where the issue comes from. This also indicates the lack of e2e testing for this important use case.

It is already mentioned the need for e2e tests at https://github.com/open-telemetry/opentelemetry-collector-contrib/issues/20552 however since that issue is more generic I would like to propose a more specific scope here to cover use cases like https://github.com/open-telemetry/opentelemetry-collector-contrib/issues/31074.

In this regard, we should consider implementing e2e tests for at least the following scenarios:

### A) Basic restart scenarios (offset tracking)

1. The collector must track file offsets, even after a restart and continue where it left off without any data loss or data duplication
2. The collector must ship logs (only once) when the backend is temporarily unavailable even if Collector restarts occur during the "backend outage" 

### B) Nice to have(s):

1. We should be able to update/downgrade/re-install without losing the offset data
2. We should correctly track offsets even in the case of a hard crash

### C) Changes on target files

1. We should be able to rotate files with move-create strategy without losing or duplicating data
2. We should be able to rotate files with copy-truncate strategy without losing or duplicating data
3. We should be able to read a file to the end even after it's deleted
4. We should be able to keep reading from a file even after it's moved (e.g: foo.log -> foo.log.1) and there should not be duplicate entries from the moved file (which could be caused due to erroneously re-reading it from the beginning)

cc: @djaglowski @ycombinator

**Comment by djaglowski** [comment]
> In this regard, we should consider implementing e2e tests for at least the following scenarios:

I agree in principle that we should have robust coverage for almost all these scenarios, but it's not clear to me whether we need these to be more than unit tests at the receiver level. For the most part, I'd like to try to find ways to localize the tests as much as possible without compromising the value they would provide. For example, I'd rather we develop a mock `consumer.Logs` which can start/stop/error at a unit test level.
 
> 1. The collector must track file offsets, even after a restart and continue where it left off without any data loss or data duplication

This is relatively well tested in unit tests. What do we gain by making this an e2e test?

> 2. The collector must ship logs (only once) when the backend is temporarily unavailable even if Collector restarts occur during the "backend outage"

We have decent coverage for non-duplication after restarting in unit tests. However, we don't have coverage which also includes backpressure from downstream consumers. This seems like a good candidate for a receiver unit test, given a mock consumer which applies backpressure.

> 1. We should be able to update/downgrade/re-install without losing the offset data

How would you test this in this repo?

> 2. We should correctly track offsets even in the case of a hard crash

This likely does require an e2e test. Do you have ideas how this would be implemented in this repo?

> 1. We should be able to rotate files with move-create strategy without losing or duplicating data
> 2. We should be able to rotate files with copy-truncate strategy without losing or duplicating data
> 4. We should be able to keep reading from a file even after it's moved and there should not be duplicate entries from the moved file (which could be caused due to erroneously re-reading it from the beginning)

These are all relatively well tested in unit tests. What do we gain by making this an e2e test?

> 3. We should be able to read a file to the end even after it's deleted

I don't think this is tested, but could be a unit test similar to the above.

**Comment by ChrsMark** [comment]
Thank's @djaglowski!

It's true that `B1` and `B2` would require some extra effort that's why I have put them as "nice-to-have"s. We can think of those in the future.

For `A1` do you mean the [testbed](https://github.com/open-telemetry/opentelemetry-collector-contrib/blob/8c943f210fa97e698f421192f25b9ab46d8e73d4/testbed/tests/log_test.go#L54-L61) or do we have another unit-test? Does this also test the correctness or just the performance part?

For `A2`, agreed.

For `C1`, `C2`, `C3`(once added) and `C4` I wonder if we should also combine the cases with a restart so as to include the persistance part as well. If that can be done with a unit-test that would be also fine. Would that make sense or a restart does not really affect these cases?

This [test-case](https://github.com/open-telemetry/opentelemetry-collector-contrib/blob/fc3966c98d6ee2490ecbfc6dcc49b1e1a61cde4f/receiver/filelogreceiver/filelog_test.go#L128-L138) covers `C1` and [this](https://github.com/open-telemetry/opentelemetry-collector-contrib/blob/fc3966c98d6ee2490ecbfc6dcc49b1e1a61cde4f/receiver/filelogreceiver/filelog_test.go#L142-L153) covers `C2`, right? I guess we can consider that `C4` is covered by `C1`, right?

**Comment by ChrsMark** [comment]
>> The collector must ship logs (only once) when the backend is temporarily unavailable even if Collector restarts occur during the "backend outage"
>
> We have decent coverage for non-duplication after restarting in unit tests. However, we don't have coverage which also includes backpressure from downstream consumers. This seems like a good candidate for a receiver unit test, given a mock consumer which applies backpressure.

Thinking of `A2` again and looking into the code (correct me if I miss anything there) I wonder if that's actually a useful unit test for the `filelog` receiver: From what I can see the `Reader` will try to send the log token to the downstream by calling the [`emitFunc/processFunc`](https://github.com/open-telemetry/opentelemetry-collector-contrib/blob/fb0c8d2cd18a2e21f7ea912cef8d1c6a83d6592e/pkg/stanza/fileconsumer/internal/reader/reader.go#L89-L93) and then it will only update the offset in case of success. Then it's downstream's responsibility to take care of it, so if the backend is unreachable then it's exporter's responsibility to persist the data properly using a [persistent-queue](https://github.com/open-telemetry/opentelemetry-collector/blob/main/exporter/exporterhelper/README.md#persistent-queue) for example. I see this is covered by [unit-tests](https://github.com/open-telemetry/opentelemetry-collector/blob/4b5d68ed282ce87beeeb2ff18e03f0d61fc285b0/exporter/exporterhelper/queue_sender_test.go#L380) which is great.
That's actually the case that I had in mind with e2e testing scenario but isolated unit-testing can also cover for now I think.


So back to your proposal for the backpressure unit test, do you mean that we could have a similar to the [TestRestartOffsets](https://github.com/open-telemetry/opentelemetry-collector-contrib/blob/ee7b5128316fd8c7a5a775ab1c9047bf20a01271/pkg/stanza/fileconsumer/file_test.go#L687) with an extra delay (==backpressure) in the [sink.Callback](https://github.com/open-telemetry/opentelemetry-collector-contrib/blob/fb87e84635f4a7e3ba1553d5d4cefce2da5b2018/pkg/stanza/fileconsumer/internal/emittest/sink.go#L60-L66)?

**Comment by crobert-1** [comment]
_It looks like there's been some good discussion here, and a general direction is agreed upon. Removing `needs triage`._

**Comment by ChrsMark** [comment]
> > From what I can see the Reader will try to send the log token to the downstream by calling the [emitFunc/processFunc](https://github.com/open-telemetry/opentelemetry-collector-contrib/blob/fb0c8d2cd18a2e21f7ea912cef8d1c6a83d6592e/pkg/stanza/fileconsumer/internal/reader/reader.go#L89-L93) and then it will only update the offset in case of success. Then it's downstream's responsibility to take care of it
> 
> I think what may be missing is in our coverage is that when the reader emits a token it is processed within `pkg/stanza/adapter` before being emitted from the receiver. We have pretty solid test coverage in `pkg/stanza/fileconsumer`, but I think `A2` would test the combination of `fileconsumer -> adapter`. This kind of test may have caught #31074 for example.

Cool, I think something like the [TestShutdownFlush](https://github.com/swiatekm-sumo/opentelemetry-collector-contrib/commit/9747d921525f2cbf4ac24727ce0d938fbb8f755c#diff-f4e11283630445cdf6387884f9753f56ae9873e1473c1b07892a9272cf397279R98) from https://github.com/swiatekm-sumo/opentelemetry-collector-contrib/commit/9747d921525f2cbf4ac24727ce0d938fbb8f755c, which tried to prove https://github.com/open-telemetry/opentelemetry-collector-contrib/issues/31074 is really close to what we need. Along with solving https://github.com/open-telemetry/opentelemetry-collector-contrib/issues/31074 we should include a test like this as well.

**Comment by ycombinator** [comment]
Reading the discussion, it seems like the pending work at this point is https://github.com/open-telemetry/opentelemetry-collector-contrib/issues/31074.  @ChrsMark @djaglowski can we close this issue here then?
